### PR TITLE
Write ArviZ docs guide

### PR DESCRIPTION
**Pull request by roshnaeem**
Description:

* Restructuring developer guide to separate documentation guide from it. 
* Updating doc guide to add more sections.

### CODE REVIEW

File: doc/source/contributing/doc_dev_summary.md
Comment (OriolAbril) [inline doc/source/contributing/doc_dev_summary.md:21]: docstrings follow numpydoc first and foremost. Some of the numpydoc constraints are checked by `pydocstyle` (it is also a check in azure), then as content of `.py` files docstrings are also subject to pylint and black rules which include those line too long and trailing whitespaces that are arguably the easiest ones to fail. But those are not the only checks nor the most important ones.

File: doc/source/contributing/doc_dev_summary.md
Comment (OriolAbril) [inline doc/source/contributing/doc_dev_summary.md:29]: very nitpicky comment. it is not github actions but readthedocs what automatically builds the docs. github actions is one of the ways to run ci checks on github, but arviz uses Azure and readthedocs

File: doc/source/contributing/docstrings.md
Comment (OriolAbril) [inline doc/source/contributing/docstrings.md:23]: The see also section, like all other sections of the docstring is already covered in the numpydoc guide. I would mention the see also guide because for us it is a key section that _must_ be present because we have a lot of interconnected functions both within the library and with external libraries like xarray for example and we have seen that users have a hard time finding the functions they are searching for. But I would not focus too much (or at all) on how to format it as numpydoc already does that, we could for example explain things that are not covered in the numpydoc style guide like how both `arviz.plot_hdi` and `plot_hdi` work and link to the same place but show different text for the hyperlink

File: doc/source/contributing/docstrings.md
Comment (OriolAbril) [inline doc/source/contributing/docstrings.md:37]: kwargs and backend_kwargs are the same aand should not be two sections. Moreover, the same reasoning used for kwargs applies to any function that calls another one. Again, I would focus **much** much more in why should things be done in a specific way and not so much on how to do those things.

File: doc/source/contributing/docstrings.md
Comment (OriolAbril) [inline doc/source/contributing/docstrings.md:40]: might be a bit of a side comment, but I think we should discuss that anyway. I envisioned this as a guide for developers/maintainers/arviz teammembers/<other alias>, with the more basic content on the above contributor guide. The steps below however are aimed at people with little to no python knowledge, I think that for the intended audience of this doc, 1-2 sentences explaining that the functions where kwargs are eventually passed to must be referenced should be enough.

File: doc/source/contributing/syntax_guide.md
Comment (OriolAbril) [inline doc/source/contributing/syntax_guide.md:56]: tabbed is from sphinx-panels and we should not use it anymore

File: doc/source/contributing/syntax_guide.md
Comment (OriolAbril) [inline doc/source/contributing/syntax_guide.md:88]: add sphinx to intersphinx mapping as it is referenced a couple times already and will probably be referenced more times

File: doc/source/contributing/syntax_guide.md
Comment (OriolAbril) [inline doc/source/contributing/syntax_guide.md:4]: doubling down on the reasoning point. It is great too highlight and show here the syntax for the more common things like anchors or toctree, however, again much more emphasis on why it's important or even imprescindible and I think all sections should have also external references. This could for example link to https://docutils.sourceforge.io/docs/ref/rst/restructuredtext.html#hyperlink-targets which extends and gives more detail on the syntax (like explaining that it can be used to link to list items or paragraphs in addition to sections) and/or https://www.sphinx-doc.org/en/master/usage/restructuredtext/roles.html#ref-role.

**Review by OriolAbril — COMMENTED** [review]
The main challenge here is defining the audience of these documents and setting their structure. I would first work on an outline detailing what you intend to cover and how are you planning to structure it, work on that at this level and then start writing and actually filling in the outline.

We also have to think about perception and biases we might end up adding on that structure. i.e. completely separating developer from doc guide can give the impression that people can work on new code features without working on documentation which is not the case, one thing is fixing an issue and consequently working only on code, but unlike when adding new doc pages which can be done on its own, adding a new feature requires writing the corresponding documentation in the same way tests need to be written too. On the other hand however, separating them can indicate that people can work on documentation without having to write code.

I am sorry a lot of the feedback is vague and non-actionable at least not without a discusion before. I am also bound to change my advise, I don't have the answers to most of the questions here. We have to experiment, try things out, discuss them and see what works best.

File: doc/source/contributing/syntax_guide.md
Comment (OriolAbril) [reply doc/source/contributing/syntax_guide.md:56]: unresolving this too. panels should not be used anymore and I removed the extension competely in #1923, it doesn't work anymore: https://arviz--1903.org.readthedocs.build/en/1903/contributing/syntax_guide.html#backticks-for-highlighting-code-keywords

File: doc/source/contributing/syntax_guide.md
Comment (OriolAbril) [reply doc/source/contributing/syntax_guide.md:4]: I don't think this is resolved. It still says linking to headings only which is the most common thing but not the only one as shown in the first reference I linked. We could consider using a seealso directive at the end of the section to add all relevant links.

File: doc/source/contributing/syntax_guide.md
Comment (OriolAbril) [inline doc/source/contributing/syntax_guide.md:74]: This should be a triple backtick block. This pre makes it look similar but it is not understood as code by sphinx. Therefore it is not modified by sphinx_copybutton for example. You can see on the preview how the copy button doesn't appear: https://arviz--1903.org.readthedocs.build/en/1903/contributing/syntax_guide.html#table-of-content-tree

File: doc/source/contributing/syntax_guide.md
Comment (OriolAbril) [inline doc/source/contributing/syntax_guide.md:88]:
```suggestion
Read more about the {ref}`Sphinx toctree directive <sphinx:table-of-contents>`.
```

ref and doc cross-references are easily ambiguous between multiple projects (unlike func ones which include the library name in the import path) and should therefore include the intersphinx key.

File: doc/source/contributing/syntax_guide.md
Comment (OriolAbril) [inline doc/source/contributing/syntax_guide.md:93]: It is also the only way for multiversioned docs to link to themselves

File: doc/source/contributing/syntax_guide.md
Comment (OriolAbril) [inline doc/source/contributing/syntax_guide.md:91]: I don't know what you thought about this section when you read it, but I think it needs to be rewritten. When I wrote it I barely understood sphinx cross-referencing and wrote all that mostly for myself. Now the separations between intersphinx or internal references, or even between ref/doc or func/meth/mod/class... ones feel forced and mostly confusing.

File: doc/source/contributing/how_to_start.md
Comment (OriolAbril) [inline doc/source/contributing/how_to_start.md:14]: These also feel more in the scope of the syntax guide than this page, and are generally covered there already.

File: doc/source/contributing/sphinx_doc_build.md
Comment (OriolAbril) [inline doc/source/contributing/sphinx_doc_build.md:2]: this document is great, I like how the sections have a couple sentence intro plus link to extended doc (which will be shared between code and doc contributions so we don't duplicate too much). The github preview section is also great, it is only missing the recommended local build instructions that I can add once we merge.

**Review by OriolAbril — CHANGES_REQUESTED** [review]
many comments reference other comments. Comments should be read all at once at least the first time before answering any or starting any changes.

File: doc/source/contributing/syntax_guide.md
Comment (OriolAbril) [reply doc/source/contributing/syntax_guide.md:115]: will do at some point in a follow up PR, or remove it and link to sphinx-primer

File: doc/source/contributing/syntax_guide.md
Comment (roshnaeem) [inline doc/source/contributing/syntax_guide.md:112]: The link is not working.

File: doc/source/contributing/syntax_guide.md
Comment (roshnaeem) [inline doc/source/contributing/syntax_guide.md:117]:
```suggestion
In ArviZ docs, we use {confval}`sphinx:intersphinx_mapping` to add references to other libraries functions and objects.
```
The link preview is supposed to be like this?

File: doc/source/contributing/diataxis_for_arviz.md
Comment (roshnaeem) [inline doc/source/contributing/diataxis_for_arviz.md:38]:
```suggestion
whereas the Example Gallery indexes by image and to avoid excessive duplication,
```

File: doc/source/contributing/architecture.md
Comment (roshnaeem) [inline doc/source/contributing/architecture.md:6]:
```suggestion
(https://github.com/arviz-devs/arviz/tree/main/arviz/stats) modules.
```

File: doc/source/contributing/architecture.md
Comment (roshnaeem) [inline doc/source/contributing/architecture.md:7]:
```
 modules.tecture
```
tecture is by mistake.

File: doc/source/contributing/syntax_guide.md
Comment (OriolAbril) [reply doc/source/contributing/syntax_guide.md:117]: no it should not. It looks like a bug in myst.

**Review by OriolAbril — APPROVED** [review]

File: doc/source/contributing/syntax_guide.md
Comment (OriolAbril) [reply doc/source/contributing/syntax_guide.md]: fixed, turns out confval cross-references are not included in sphinx, they must be added manually if we want to reference them.

Comment (roshnaeem) [reply doc/source/contributing/syntax_guide.md]: yes, it is working fine now.

**Pull request merged**
